File: packages/node/src/RegexDetector.ts
import { Detector, DetectorResult } from "./types";

const EMAIL_RE = /[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}/g;
const PHONE_RE = /\+?\d[\d\-\s]{7,}\d/g;
const NAME_RE = /\b[A-Z][a-z]+(?:\s+[A-Z][a-z]+)+\b/g;
const IP_RE = /\b(?:\d{1,3}\.){3}\d{1,3}\b/g;
const CC_RE = /\b\d{4}[\s\-]?\d{4}[\s\-]?\d{4}[\s\-]?\d{4}\b/g;
const SSN_RE = /\b\d{3}-\d{2}-\d{4}\b/g;

/**
 * Simple regex-based PII detector for Node/TypeScript.
 *
 * Detects:
 * - EMAIL: Email addresses
 * - PHONE: Phone numbers
 * - PERSON: Names (simple capitalized word patterns)
 * - IP_ADDRESS: IPv4 addresses
 * - CREDIT_CARD: Credit card numbers
 * - SSN: Social Security Numbers
 */
export class RegexDetector implements Detector {
  detect(text: string): DetectorResult[] {
    const results: DetectorResult[] = [];

    // Helper function to find all matches
    const findMatches = (regex: RegExp, entityType: string) => {
      // Reset regex lastIndex
      regex.lastIndex = 0;
      let match: RegExpExecArray | null;
      while ((match = regex.exec(text)) !== null) {
        results.push({
          entityType,
          start: match.index,
          end: match.index + match[0].length,
          text: match[0],
        });
      }
    };

    findMatches(EMAIL_RE, "EMAIL");
    findMatches(PHONE_RE, "PHONE");
    findMatches(NAME_RE, "PERSON");
    findMatches(IP_RE, "IP_ADDRESS");
    findMatches(CC_RE, "CREDIT_CARD");
    findMatches(SSN_RE, "SSN");

    return results;
  }
}

File: packages/python/src/prompt_guard/detectors/regex_detector.py
import re
from typing import List
from .base import BaseDetector
from ..types import DetectorResult

EMAIL_RE = re.compile(r"[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}")
PHONE_RE = re.compile(r"\+?\d[\d\-\s]{7,}\d")
# Simple name pattern - detects capitalized words that look like names
NAME_RE = re.compile(r"\b[A-Z][a-z]+(?:\s+[A-Z][a-z]+)+\b")
# IP address pattern
IP_RE = re.compile(r"\b(?:\d{1,3}\.){3}\d{1,3}\b")
# Credit card pattern (simple)
CC_RE = re.compile(r"\b\d{4}[\s\-]?\d{4}[\s\-]?\d{4}[\s\-]?\d{4}\b")
# SSN pattern
SSN_RE = re.compile(r"\b\d{3}-\d{2}-\d{4}\b")


class RegexDetector(BaseDetector):
    """
    Simple regex-based PII detector.

    Detects:
    - EMAIL: Email addresses
    - PHONE: Phone numbers
    - PERSON: Names (simple capitalized word patterns)
    - IP_ADDRESS: IPv4 addresses
    - CREDIT_CARD: Credit card numbers
    - SSN: Social Security Numbers
    """

    def detect(self, text: str) -> List[DetectorResult]:
        results: List[DetectorResult] = []

        # Detect emails
        for match in EMAIL_RE.finditer(text):
            results.append(
                DetectorResult(
                    entity_type="EMAIL",
                    start=match.start(),
                    end=match.end(),
                    text=match.group(0),
                )
            )

        # Detect phone numbers
        for match in PHONE_RE.finditer(text):
            results.append(
                DetectorResult(
                    entity_type="PHONE",
                    start=match.start(),
                    end=match.end(),
                    text=match.group(0),
                )
            )

        # Detect names
        for match in NAME_RE.finditer(text):
            results.append(
                DetectorResult(
                    entity_type="PERSON",
                    start=match.start(),
                    end=match.end(),
                    text=match.group(0),
                )
            )

        # Detect IP addresses
        for match in IP_RE.finditer(text):
            results.append(
                DetectorResult(
                    entity_type="IP_ADDRESS",
                    start=match.start(),
                    end=match.end(),
                    text=match.group(0),
                )
            )

        # Detect credit cards
        for match in CC_RE.finditer(text):
            results.append(
                DetectorResult(
                    entity_type="CREDIT_CARD",
                    start=match.start(),
                    end=match.end(),
                    text=match.group(0),
                )
            )

        # Detect SSNs
        for match in SSN_RE.finditer(text):
            results.append(
                DetectorResult(
                    entity_type="SSN",
                    start=match.start(),
                    end=match.end(),
                    text=match.group(0),
                )
            )

        return results
